File: src/include/asset/Asset.php
<?php

namespace asset;

use creator\Creator;
use DateTime;

/**
 * The main asset class.
 * It represents one PBR material, 3D model or other asset.
 */
abstract class Asset
{
	public function __construct(
		public ?int $id,
		public ?string $creatorGivenId,
		public string $title,
		public string $url,
		public DateTime $date,
		public AssetType $type,
		public Creator $creator,
		public array $tags = [],
	) {}
}

File: app/asset/asset.py
from dataclasses import dataclass, field
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from app.asset.asset_type import AssetType
    from app.creator.creator import Creator


@dataclass
class Asset:
    id: Optional[int]
    creator_given_id: Optional[str]
    title: str
    url: str
    type: "AssetType"
    creator: "Creator"
    tags: list[str] = field(default_factory=list)
